### PR TITLE
[Warmup] Tune the LM head GEMM during FlashInfer autotune

### PR DESCRIPTION
**Pull request by askliar**
## Summary

FlashInfer autotune warmup only runs `_dummy_run`, which executes the transformer body but never calls `compute_logits`. As a result the LM head (hidden→vocab) GEMM is never observed by the autotuner, and at runtime the first decode batch triggers:
```
[AutoTuner]: No tuned config covers bf16_gemm input_shapes=(... [H, vocab] ...)
```
falling back to an untuned heuristic tactic for a GEMM that runs every decode step.

## Change
`flashinfer_autotune` now routes all three `_dummy_run` call sites (the non-persistent-cache branch and both leader/non-leader persistent-cache paths) through a helper that also runs `_dummy_sampler_run` on the returned hidden states, guarded by `is_last_rank` and `not is_pooling_model`. This mirrors how `profile_run` exercises the LM head, so the projection GEMM is tuned inside the `autotune()` context. Calling it on both leader/non-leader preserves collective (TP all-gather) symmetry.

## Test plan
- [ ] Serve a BF16 model with `--enable-flashinfer-autotune` and confirm the
      `No tuned config covers bf16_gemm` warning for the `[*, hidden] x
      [hidden, vocab]` shape no longer appears at runtime.
- [ ] Verify autotune warmup completes without regressions on a TP>1 run.

## Duplicate-work check
- None open

## AI assistance
AI assistance was used (Claude) to write this PR

### CODE REVIEW

**Comment by jdmays13** [comment]
**Verified on GB10 / DGX Spark (sm_121): the patch achieves its stated goal, but throughput was lower in my testing on an NVFP4 LM head.**

I back-ported this onto v0.25.0 and tested it on a `Qwen3.8-27B` NVFP4 checkpoint — vocab 248,320, `lm_head` quantised to NVFP4, MTP speculative decoding with `num_speculative_tokens=4`. Note the warning in this configuration is `fp4_gemm`, not the `bf16_gemm` in your description.

### Test plan item 1: confirmed

Before the patch, every boot logged three fallback warnings for the `(m, 2560) x (2560, 248320)` GEMM at **m=1** (single-token decode), **m=5** (K+1 verify width) and **m=16** (`max_num_seqs`). After the patch: **zero**. The LM head GEMM is genuinely being tuned, at all the widths that matter at runtime.

### However, throughput was consistently lower with the patch

| arm | median t/s (median-of-7, warm-up discarded) | `No tuned config covers ...248320` |
|---|---|---|
| unpatched, fresh tuning cache | **28.7** (28.7–28.8) | 3 — m=[1, 5, 16] |
| this PR, fresh tuning cache | **26.5** (26.3–26.6) | 0 |

**Caveat on the magnitude.** On this box the tuning pass itself is not deterministic: the same image and config with a *different* autotune cache measured 30.1 vs the 28.7 above, i.e. ~5% spread attributable to the tuning run alone. The patch penalty here (7.7%) exceeds that, but it comes from a single paired run — so please treat the *direction* as the finding and the *magnitude* as indicative rather than precise.

### Method

Identical model, flags and `num_speculative_tokens` in both arms; only the image and the autotune cache directory differed. Each arm ran with a freshly wiped `VLLM_FLASHINFER_AUTOTUNE_CACHE_DIR` so both performed a real tuning pass, making the patch the only variable. Both arms ran back-to-back. For reference, within-boot measurement spread on this hardware is ±0.2 t/s, while across-boot drift on an identical config can reach ~14%, which is why every number above is a median-of-7 with a warm-up discarded.

### Interpretation

On sm_121 with an NVFP4 `lm_head`, the tuned tactic FlashInfer selects for this GEMM appears **no better, and possibly worse**, than the heuristic it replaces. In this configuration the warning may be a false alarm rather than a missed optimisation.

I can't distinguish whether that is specific to sm_121, to FP4 weights, or to very large vocabularies from a single machine. Happy to run further variants if useful — an FP8 `lm_head` build of the same model is available locally, which would separate the FP4 question from the large-vocab one.

### One additional observation

The shapes needing coverage here (m=1 and m=K+1) are driven by speculative decoding and are *narrower* than the profile batch width the tuning pass runs at. Coverage worked out fine in my run, but it may be worth confirming on setups without speculative decoding, where the decode width is only m=1.

*Tested with AI assistance (Claude). All figures above are measured on real hardware, not estimated.*

**Comment by webcodes-cz** [comment]
Tested the PR mechanism as a minimal backport of head
`7b73de0d43d5dba2efc2fc80a9af3223df3fbb90` onto vLLM 0.27.1
(`6e448d0ea9bf3d88d898b65449ca6dc2aec170ac`) on an RTX 5090 (sm_120),
using Qwen3.8-27B NVFP4 with a 248,320-token vocabulary and an NVFP4
`lm_head`. MTP and speculative decoding were disabled.

## Method

- Exact same checkpoint in both arms: revision
  `0cc27958cefbbe231782ec8511de8c4eb5233348`, 18,765,214,312 indexed bytes
  in three shards.
- Exact base image versus a derived image in which only `kernel_warmup.py` and
  provenance metadata changed.
- Separate fresh FlashInfer cache roots for the two arms.
- Seven predeclared paired repetitions at C1, C8, C16, and C24; 64 random
  input tokens, exactly 512 output tokens, one unscored C8 warmup, and no
  preemption.
- Pair order and selection gate were frozen before execution.

## Mechanism

The unpatched arm produced 80 FP4 cache records, zero `lm_head` records, and
five fallback warnings. The patched arm produced 87 FP4 records, seven
`lm_head` records, and zero fallback warnings. The added `lm_head` rows covered
M = 1, 2, 4, 8, 16, 32, and 64.

## Paired output-throughput result

| Concurrency | Median paired gain | Range across 7 pairs |
| ---: | ---: | ---: |
| C1 | +1.788% | +1.620% to +2.031% |
| C8 | +1.377% | +0.976% to +1.810% |
| C16 | +1.065% | +0.717% to +1.723% |
| C24 | +0.502% | +0.300% to +0.639% |

All 28 paired observations were positive. The maximum within-arm spread was
0.840%. On sm_120 without speculative decoding, this is therefore a small but
consistent throughput improvement, rather than a large performance cliff.

We also ran a separate temperature-zero, seed-38, 40-case C8 quality A-B-A.
That result was inconclusive: the two identical unpatched boots differed on
28/40 normalized responses, more than either comparison involving the patched
arm. Aggregate scores for the patched arm matched unpatched A, while
unpatched B moved legal/exact/incomplete counts. We therefore make no
exact-output quality-equivalence or regression claim from that generation
canary.

## AI assistance

AI assistance was used (Codex and Claude) to design and review the frozen
harness and audit the receipts. Measurements were executed on real hardware.
